README.md
# Meta Async Programming Approach

[![Build Status](https://travis-ci.org/metarhia/MetaSync.svg?branch=master)](https://travis-ci.org/metarhia/MetaSync)
[![Dependency Status](https://david-dm.org/metarhia/MetaSync.svg)](https://david-dm.org/metarhia/MetaSync)

## Installation

```bash
$ npm install metasync
```

# Examples

## Functional Asyncronous Composition

Syntax: `metasync.composition(functions, [done, [data]]);`

Parameters:
- functions - array of `function([data,] callback)` where:
  - data - optional incoming data
  - callback - `function([data])` where:
    - data - outgoing data
- done - `callback(data)` where:
  - data - hash with of functions results
- data - incoming data

```JavaScript
metasync.composition(
  [f1, f2, f3, [[f4, f5, [f6, f7], f8]], f9]
);
```

- Array of functions gives sequential execution: `[f1, f2, f3]`
- Double brackets array of functions gives parallel execution: `[[f1, f2, f3]]`

Examples of functions:

```JavaScript
// with one argiment and no result
function f1(callback) {
  callback();
}

// with one argiment and returning result
function f2(callback) {
  callback('value');
}

// with two argiments and result
function f3(data, callback) {
  // assign result to data random key
  data.keyName = 'value';
  // returning no result
  callback();
}
```

## An Event-driven Asyncronous Data Collector

```JavaScript
var metasync = require('metasync');
var fs = require('fs');

var dataCollector = new metasync.DataCollector(4, (data) => {
  console.dir(Object.keys(data));
});

dataCollector.collect('user', { name: 'Marcus Aurelius' });

fs.readFile('HISTORY.md', (err, data) => {
  dataCollector.collect('history', data);
});

fs.readFile('README.md', (err, data) => {
  dataCollector.collect('readme', data);
});

setTimeout(() => {
  dataCollector.collect('timer', { date: new Date() });
}, 1000);
```

## Parallel execution

```JavaScript
metasync.parallel([f1, f2, f3], () =>  { ... });
```

## Sequential execution

```JavaScript
metasync.sequential([f1, f2, f3], () => { ... });
```

## Asynchrous filter

```JavaScript
metasync.filter(['data', 'to', 'filter'], (item, callback) => {
  callback(item.length > 2);
}, (result) => {
  console.dir(result);
});
```

## Asynchrous find

```JavaScript
metasync.find(
  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16],
  (item, callback) => (
    callback(item % 3 === 0 && item % 5 === 0)
  ),
  (result) => {
    console.dir(result);
  }
);
```

## Asyncronous series (sequential)

```JavaScript
metasync.series(
  ['a', 'b', 'c'],
  (item, callback) => {
    console.dir({ series: item });
    callback();
  },
  (data) => {
    console.dir('series done');
  }
);
```

## Asyncronous each (parallel)

```JavaScript
metasync.each(
  ['a', 'b', 'c'],
  (item, callback) => {
    console.dir({ each: item });
    callback();
  },
  (data) => {
    console.dir('each done');
  }
);
```

## Contributors

  - Timur Shemsedinov (marcusaurelius)
  - See github for full [contributors list](https://github.com/metarhia/MetaSync/graphs/contributors)


HISTORY.md
0.1.1 / 2017-01-01
==================

  * Code style, issue #33

0.1.0 / 2016-12-31
==================

  * Refactored to ES6
  * Removed old Node.js support, now MetaSync will support just Node.js 6.x and 7.x

0.0.19 / 2016-12-11
==================

  * Code style

0.0.18 / 2016-12-05
==================

  * Fixed clearTimeout usage, close #25

0.0.17 / 2016-12-01
==================

  * Added arguments to metasync.throttle, issue #20

0.0.16 / 2016-11-30
==================

  * Added final call to metasync.throttle after last function call

0.0.15 / 2016-11-30
==================

  * Implementation of function throttling, close #20

0.0.14 / 2016-08-25
==================

  * ConcurrentQueue fixed

0.0.13 / 2016-08-24
==================

  * Implemented ConcurrentQueue

0.0.12 / 2016-08-23
==================

  * Fixed DataCollector

0.0.11 / 2016-08-23
==================

  * Removed DataCollector.on('collect') event
  * Added errors hash to DataCollector.on('done', function(errs, data))

0.0.10 / 2016-08-23
==================

  * Changed KeyCollector interface
  * Changed ConcurentQueue interface
  * Removed debug output

0.0.9 / 2016-08-21
==================

  * Implemented async reduce
  * Multiple fixes

0.0.8 / 2016-08-17
==================

  * Fixed error returning
  * Removed unused arguments in tests

0.0.7 / 2016-08-14
==================

  * Added data hash for metasync.composition()
  * Added tests and examples
  * Restored done for metasync.composition()

0.0.6 / 2016-08-13
==================

  * Added optional done in metasync.each
  * Fixed another critical bug with done
  * Removed done for metasync.composition()

0.0.5 / 2016-08-03
==================

  * Fixed critical bug, done function is optional now

0.0.4 / 2016-07-30
==================

  * Implemented metasync.each
  * Implemented metasync.series
  * Code style fixes and optimizations

0.0.3 / 2016-07-27
==================

  * Added Trevis CI tests
  * Implemented metasync.filter
  * Implemented metasync.find
  * Code style fixes and optimizations

0.0.2 / 2016-07-23
==================

  * Implemented metasync.composition, issue #1
  * Implemented metasync.parallel
  * Implemented metasync.sequential

0.0.1 / 2016-07-18
==================

  * An event-driven asyncronous data collector implemented: metasync.DataCollector


examples.js
'use strict';

const metasync = require('./metasync');
const fs = require('fs');
const assert = require('assert');
const ASYNC_TIMEOUT = 200;

// assert.deepStrictEqual polyfill for pre-io.js
if (!assert.deepStrictEqual) {
  assert.deepStrictEqual = (actual, expected, message) => {
    const actualKeys = Object.keys(actual);
    const expectedKeys = Object.keys(expected);
    assert.strictEqual(actualKeys.length, expectedKeys.length, message);
    let i, key;
    for (i = 0; i < actualKeys.length; i++) {
      key = actualKeys[i];
      assert.strictEqual(actual[key], expected[key], message);
    }
  };
}

// Functional Asyncronous Composition

function compositionTest(end) {

  metasync.composition(
    [f1, f2, f3, [[f4, f5, [f6, f7], f8]], [[f9, f10]], f11],
    (data) => console.dir(data)
  );

  function f1(callback) {
    console.log('f1');
    setTimeout(() => callback('result1'), ASYNC_TIMEOUT);
  }

  function f2(data, callback) {
    console.log('f2');
    setTimeout(() => callback('result2'), ASYNC_TIMEOUT);
  }

  function f3(data, callback) {
    console.log('f3');
    setTimeout(() => {
      data.third = 'result3';
      callback();
    }, ASYNC_TIMEOUT);
  }

  function f4(callback) {
    console.log('f4');
    setTimeout(() => callback(), ASYNC_TIMEOUT);
  }

  function f5(data, callback) {
    console.log('f5');
    setTimeout(() => callback(5), ASYNC_TIMEOUT);
  }

  function f6(data, callback) {
    console.log('f6');
    setTimeout(() => callback('result6'), ASYNC_TIMEOUT);
  }

  function f7(data, callback) {
    console.log('f7');
    setTimeout(() => callback('result7'), ASYNC_TIMEOUT);
  }

  function f8(data, callback) {
    console.log('f8');
    setTimeout(() => callback('result8'), ASYNC_TIMEOUT);
  }

  function f9(data, callback) {
    console.log('f9');
    setTimeout(() => callback('result9'), ASYNC_TIMEOUT);
  }

  function f10(data, callback) {
    console.log('f10');
    setTimeout(() => callback('result10'), ASYNC_TIMEOUT);
  }

  function f11(data, callback) {
    console.log('f11');
    setTimeout(() => {
      callback('result11');
      console.log('Composition test done');
      end();
    }, ASYNC_TIMEOUT);
  }

}

// Data Collector

function dataCollectorTest(end) {

  const dataCollector = new metasync.DataCollector(4);

  dataCollector.on('done', (errs, data) => {
    console.dir({
      dataKeys: Object.keys(data)
    });
    console.log('DataCollector test done');
    end();
  });

  dataCollector.collect('user', { name: 'Marcus Aurelius' });

  fs.readFile('HISTORY.md', (err, data) => {
    dataCollector.collect('history', data);
  });

  fs.readFile('README.md', (err, data) => {
    dataCollector.collect('readme', data);
  });

  setTimeout(() => {
    dataCollector.collect('timer', { date: new Date() });
  }, ASYNC_TIMEOUT);

}

function dataCollectorTimeoutTest(end) {

  const dataCollector = new metasync.DataCollector(4, 1000);

  dataCollector.on('timeout', (err, data) => {
    console.dir({
      err: err ? err.toString() : null,
      dataKeys: Object.keys(data)
    });
    console.log('Collector timeout test done');
    end();
  });

  dataCollector.collect('user', { name: 'Marcus Aurelius' });

}

function dataCollectorErrorTest(end) {

  const dataCollector = new metasync.DataCollector(4);

  dataCollector.on('error', (err, key) => {
    console.dir({
      err: err ? err.toString() : null,
      key
    });
  });

  dataCollector.on('done', (errs, data) => {
    console.dir({
      errorKeys: errs ? Object.keys(errs) : null,
      dataKeys: Object.keys(data)
    });
    console.log('Collector Error test done');
    end();
  });

  dataCollector.collect('user', new Error('User not found'));
  dataCollector.collect('file', 'file content');
  dataCollector.collect('score', 1000);
  dataCollector.collect('tcp', new Error('No socket'));

}

// Key Collector

function keyCollectorTest(end) {

  const keyCollector = new metasync.KeyCollector(['user', 'history'], 1000);

  keyCollector.on('done', (errs, data) => {
    console.dir({
      dataKeys: Object.keys(data)
    });
    console.log('KeyCollector test done');
    end();
  });

  keyCollector.collect('user', { name: 'Marcus Aurelius' });

  fs.readFile('HISTORY.md', (err, data) => {
    keyCollector.collect('history', data);
  });

}
// Parallel execution

function parallelTest(end) {

  metasync.parallel([pf1, pf2, pf3], () => {
    console.log('Parallel test done');
    end();
  });

  function pf1(data, callback) {
    console.log('pf1');
    setTimeout(() => callback('result1'), ASYNC_TIMEOUT);
  }

  function pf2(data, callback) {
    console.log('pf2');
    setTimeout(() => callback('result2'), ASYNC_TIMEOUT);
  }

  function pf3(data, callback) {
    console.log('pf3');
    setTimeout(() => callback('result3'), ASYNC_TIMEOUT);
  }

}

// Sequential execution

function sequentialTest(end) {

  metasync.sequential([sf1, sf2, sf3], () => {
    console.log('Sequential test done');
    end();
  });

  function sf1(data, callback) {
    console.log('sf1');
    setTimeout(() => callback('result1'), ASYNC_TIMEOUT);
  }

  function sf2(data, callback) {
    console.log('sf2');
    setTimeout(() => callback('result2'), ASYNC_TIMEOUT);
  }

  function sf3(data, callback) {
    console.log('sf3');
    setTimeout(() => callback('result3'), ASYNC_TIMEOUT);
  }

}

// Asynchrous filter

function filterTest(end) {

  const dataToFilter = [
    'Lorem', 'ipsum', 'dolor', 'sit', 'amet', 'consectetur',
    'adipiscing', 'elit', 'sed', 'do', 'eiusmod', 'tempor',
    'incididunt', 'ut', 'labore', 'et', 'dolore', 'magna', 'aliqua',
  ];

  function filterPredicate(item, callback) {
    // filter words which consists of unique letters only
    const letters = [];
    console.log('checking value: ' + item);
    let i;
    for (i = 0; i < item.length; ++i) {
      if (letters.includes(item[i].toLowerCase())) {
        break;
      }
      letters.push(item[i].toLowerCase());
    }

    setTimeout(() => callback(letters.length === item.length), ASYNC_TIMEOUT);
  }

  metasync.filter(dataToFilter, filterPredicate, (result) => {
    console.log('filtered array: ' + result);
    console.log('Filter test done');
    end();
  });

}

// Asynchrous find

function findTest(end) {

  metasync.find(
    [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16],
    (item, callback) => (
      callback(item % 3 === 0 && item % 5 === 0)
    ),
    (result) => {
      console.log('found value: ' + result);
      console.log('Find test done');
      end();
    }
  );

}

// Asyncronous each in parallel

function eachTest(end) {

  metasync.each(
    ['a', 'b', 'c'],
    (item, callback) => {
      console.dir({ each: item });
      callback();
    },
    (/*data*/) => {
      console.log('Each test done');
      end();
    }
  );

}

// Asyncronous series (sequential)

function seriesTest(end) {

  metasync.series(
    ['a', 'b', 'c'],
    (item, callback) => {
      console.dir({ series: item });
      callback();
    },
    (/*data*/) => {
      console.log('Series test done');
      end();
    }
  );

}

// Asyncronous reduce (sequential)

function reduceTest(end) {

  metasync.reduce(
    ['a', 'b', 'c'],
    (prev, curr, callback) => {
      console.dir({ reduce: { prev, curr } });
      callback(null, curr);
    },
    (/*err, data*/) => {
      console.log('Reduce test done');
      end();
    }
  );

}

function concurrentQueueTest(end) {

  const queue =  new metasync.ConcurrentQueue(3, 2000);

  queue.on('process', (item, callback) => {
    setTimeout(() => {
      console.dir({ item });
      callback();
    }, 100);
  });

  queue.on('timeout', () => {
    console.log('ConcurrentQueue timed out');
  });

  queue.on('empty', () => {
    console.log('ConcurrentQueue test done');
    end();
  });

  queue.add({ id: 1 });
  queue.add({ id: 2 });
  queue.add({ id: 3 });
  queue.add({ id: 4 });
  queue.add({ id: 5 });
  queue.add({ id: 6 });
  queue.add({ id: 8 });
  queue.add({ id: 9 });
}

function concurrentQueuePauseResumeStopTest(end) {
  const queue =  new metasync.ConcurrentQueue(3, 2000);
  queue.pause();
  queue.on('empty', end);
  if (!queue.events.empty) {
    console.log('ConcurrentQueue pause test done');
  }
  queue.resume();
  queue.on('empty', end);
  if (queue.events.empty) {
    console.log('ConcurrentQueue resume test done');
  }
  queue.stop();
  if (queue.count === 0) {
    console.log('ConcurrentQueue stop test done');
  }
}

function throttleTest(end) {

  let state;

  function fn(letter) {
    console.log('Throttled function, state: ' + state);
    if (state === letter) {
      console.log('Throttle test done');
      end();
    }
  }

  const f1 = metasync.throttle(500, fn, ['I']);

  // to be called 2 times (first and last: A and E)
  state = 'A';
  f1();
  state = 'B';
  f1();
  state = 'C';
  f1();
  state = 'D';
  f1();
  state = 'E';
  f1();

  // to be called 2 times (last will be I)
  setTimeout(() => {
    state = 'F';
    f1();
  }, 600);
  setTimeout(() => {
    state = 'G';
    f1();
  }, 700);
  setTimeout(() => {
    state = 'H';
    f1();
  }, 1000);
  setTimeout(() => {
    state = 'I';
    f1();
  }, 1100);

}

function mapTest() {
  metasync.map([1, 2, 3], (item, callback) => {
    setTimeout(() => {
      callback(null, item * item);
    }, item * 10);
  }, (error, result) => {
    assert.ifError(error);
    assert.deepStrictEqual(result, [1, 4, 9]);
    console.log('Map test #1 done');
  });

  metasync.map([1, 2, 3], (item, callback) => {
    setTimeout(() => {
      if (item === 2) {
        callback(new Error());
      } else {
        callback(null, item);
      }
    }, item * 10);
  }, (error, result) => {
    assert.ok(error);
    assert.ifError(result);
    console.log('Map test #2 done');
  });
}

function timeoutTest() {
  // Done function called by timer
  const start1 = new Date();
  metasync.timeout(200, (done) => {
    setTimeout(done, 300);
  }, () => {
    const timeDiff = new Date() - start1;
    assert(timeDiff < 250);
    console.log('Timout test #1 done');
  });

  // Done function called by async function
  const start2 = new Date();
  metasync.timeout(300, done => setTimeout(done, 200), () => {
    const timeDiff = new Date() - start2;
    assert(timeDiff < 250);
    console.log('Timout test #2 done');
  });
}

// Run tests

metasync.composition([
  compositionTest,
  dataCollectorTest,
  dataCollectorTimeoutTest,
  dataCollectorErrorTest,
  keyCollectorTest,
  parallelTest,
  sequentialTest,
  filterTest,
  findTest,
  eachTest,
  seriesTest,
  reduceTest,
  concurrentQueueTest,
  concurrentQueuePauseResumeStopTest,
  throttleTest,
  mapTest,
  timeoutTest
], () => {
  console.log('All tests done');
});
